File: dev/python/drugbank_parse/schema.py
from __future__ import annotations

from pathlib import Path

import yaml

from .models import DrugBankSchema, TableSchema


def default_schema_dir() -> Path:
    return Path(__file__).resolve().parents[2] / "schema"


def load_schema(schema_dir: str | Path | None = None) -> DrugBankSchema:
    base = Path(schema_dir) if schema_dir is not None else default_schema_dir()
    tables_data = _read_yaml(base / "tables.yml")
    fields_data = _read_yaml(base / "fields.yml")

    tables = {}
    for name, definition in tables_data["tables"].items():
        tables[name] = TableSchema(
            name=name,
            description=definition.get("description", ""),
            columns=list(definition["columns"]),
            required=list(definition.get("required", [])),
        )

    return DrugBankSchema(
        version=int(tables_data["version"]),
        tables=tables,
        fields=dict(fields_data.get("fields", {})),
    )


def _read_yaml(path: Path) -> dict:
    if not path.exists():
        raise FileNotFoundError(f"Schema file does not exist: {path}")

    with path.open("r", encoding="utf-8") as handle:
        data = yaml.safe_load(handle)

    if not isinstance(data, dict):
        raise ValueError(f"Schema file is empty or invalid: {path}")
    return data

File: dev/R/R/schema.R
load_yaml_file <- function(path) {
  if (!file.exists(path)) {
    stop(sprintf("Schema file does not exist: %s", path), call. = FALSE)
  }
  value <- yaml::read_yaml(path)
  if (!is.list(value)) {
    stop(sprintf("Schema file is empty or invalid: %s", path), call. = FALSE)
  }
  value
}

load_drugbank_schema <- function(schema_dir = NULL) {
  base <- if (is.null(schema_dir)) default_schema_dir() else schema_dir
  tables <- load_yaml_file(file.path(base, "tables.yml"))
  fields <- load_yaml_file(file.path(base, "fields.yml"))

  list(
    version = as.integer(tables$version),
    tables = tables$tables,
    fields = fields$fields
  )
}

load_drugbank_profiles <- function(schema_dir = NULL) {
  base <- if (is.null(schema_dir)) default_schema_dir() else schema_dir
  load_yaml_file(file.path(base, "profiles.yml"))
}

resolve_modules <- function(profile = "core", modules = NULL, schema_dir = NULL) {
  profiles <- load_drugbank_profiles(schema_dir)
  known_profiles <- profiles$profiles
  known_modules <- profiles$modules

  if (!profile %in% names(known_profiles)) {
    stop(sprintf("Unknown profile: %s", profile), call. = FALSE)
  }

  selected <- if (is.null(modules)) known_profiles[[profile]]$modules else modules
  for (module in selected) {
    if (!module %in% names(known_modules)) {
      stop(sprintf("Unknown module: %s", module), call. = FALSE)
    }
  }
  selected
}

resolve_tables <- function(modules, schema_dir = NULL) {
  profiles <- load_drugbank_profiles(schema_dir)
  known_modules <- profiles$modules
  tables <- character()

  for (module in modules) {
    if (!module %in% names(known_modules)) {
      stop(sprintf("Unknown module: %s", module), call. = FALSE)
    }
    for (table in known_modules[[module]]$tables) {
      if (!table %in% tables) {
        tables <- c(tables, table)
      }
    }
  }
  tables
}
